### ISSUE TITLE
React@16.9 block `javascript:void(0);`

### ISSUE THREAD

**Issue by fengxinming** [facebook/react#16592]
```
<a href="javascript:void(0);"></a>
```

```
index.js:1375 Warning: A future version of React will block javascript: URLs as a security precaution. Use event handlers instead if you can. If you need to generate unsafe HTML try using dangerouslySetInnerHTML instead. React was passed "javascript:void(0);".

```
react@16.9

**Comment by fengxinming** [comment]
@aweary please you provide us a solution to solve the change you made!

**Comment by pllee** [comment]
@aweary is there a React recommended way to get the previous behavior of `href="javascript:void(0);"`?  Not having an href gets rid of the default keyboard behavior tab and tab -> enter to behave like `onClick`.  Using `#` gets that behavior but the `onClick` handler then has the onus of stopping the event propagation to not get the default browser behavior of scrolling to the top.

**Comment by hamlim** [comment]
You could use `<button type="button" onClick={}>...</button>`, the added benefit here is that the role of the element is better communicated to screen reader users using your feature 👍

**Comment by JavaJamie** [comment]
We encountered this issue too. We were already using the react-router-dom library so opted to use its NavLink which resolved the issue and it perfectly mimics the javascript:void(0) behaviour (but without the warning!).

`<NavLink to="#" onClick={e => e.preventDefault()}>`

**Comment by raxod502** [comment]
@aweary If the ability to use `javascript:void(0)` is being removed then React should ship an alternative way to create links that act like buttons. [A brief glance at StackOverflow](https://stackoverflow.com/a/1368286/3538165) shows that there is no reasonable alternative to `javascript:void(0)`. What about just whitelisting this specific string? That would solve this usability problem while not introducing any security concerns.

**Comment by dondmcg** [comment]
@aweary I agree with raxod502 I have a specific problem where I am using an API for authentication. If I change the element from a anchor to a button the API no longer works because the js that is not exposed to me is still looking for an a tag. I will be forced to put inline js for prevent default and change the href to a hash tag which I do not think is a good solution.

**Comment by gaearon** [comment]
See my reply in https://github.com/facebook/react/issues/16382#issuecomment-607252694.

**Comment by raxod502** [comment]
When I posted my original comment, I was under the mistaken impression that you could not work around the problem by setting `href="#"` and adding `event.preventDefault()` to the `onClick` handler. (I must have tested incorrectly.)

Could this be prominently documented? Finding out about it deep in the issue tracker is a very bad user experience.

**Comment by mythz** [comment]
This change is unnecessarily user hostile. 

What's the rationale for not whitelisting `href="javascript:"` or `href="javascript:void(0)"` so everyone can have a minimally disruptive uncompromising solution?

**Comment by TomerHeber** [comment]
This is how I worked around it (created a new component):
(A bit more verbose because I'm using typescript).

```
import * as React from "react";

interface OnClickLinkProps {
  children: React.ReactChild | React.ReactChild[];
  className?: string;
  onClick: () => void;
}

export function OnClickLink(props: OnClickLinkProps): React.ReactElement {
  function onClick(event: React.MouseEvent<HTMLAnchorElement>) {
    event.preventDefault();
    props.onClick();
  }

  return (
    <a href="#" onClick={onClick} className={props.className || ""}>
      {props.children}
    </a>
  );
}
```

**Comment by ksenginew** [comment]
`href` attribute without value and equal sign works
```html
<a href></a>
```
To make sure its cross browser compatible use unused ID.
```html
<a href="#nolink"></a>
```
> Don't add `id="nolink"` to any tag in HTML page.

**Comment by lynda0214** [comment]
> ```
> href
> ```

The first one would throw a warning as follows:
Received `true` for a non-boolean attribute `href`

**Comment by dancherb** [comment]
This warning is coming up for me using next.js's next/link tag, which requires a value for href, even when simply using it for onPress. None of the suggestions so far have been appropriate as a replacement for javascript:void(0).

Passing a href value and using it with e.preventDefault() isn't appropriate because there may be other effects you do want to capture (e.g. a dropdown registering the click and closing itself).

**Comment by wmzy** [comment]
The following method can avoid React errors:
```jsx
<a href={maybeEmpty ?? 'blob:'>...</a>
```
The disadvantage is that clicking it will trigger a console error.
